### PR TITLE
Add VisuShrink method for `denoise_wavelet`

### PR DESCRIPTION
**Pull request by monego**
## Description

This PR extends `denoise_wavelet` with the VisuShrink thresholding method [1].

## Checklist

- [x] Gallery example in `./doc/examples` (new features only)
- [x] Unit tests

## References

[1] DL Donoho and JM Johnstone, Ideal spatial adaptation by wavelet shrinkage, Biometrika, 1994, vol. 81, no. 3, pp. 425-455.

### CODE REVIEW

**Comment by sciunto** [comment]
@monego thanks for your contribution. Is it still a work in progress? Could you explain in few words the differences between these two methods in practice, maybe with an example?

File: skimage/restoration/_denoise.py [hunk 1]
```diff
@@ -349,6 +349,10 @@ def _bayes_thresh(details, var):
     thresh = var / np.sqrt(max(dvar - var, eps))
     return thresh
 
+def _universal_thresh(img, details, sigma):
```
Comment (grlee77) [inline skimage/restoration/_denoise.py]: why pass in details when it is not used?

File: skimage/restoration/_denoise.py [hunk 2]
```diff
@@ -459,8 +466,13 @@ def _wavelet_threshold(img, wavelet, threshold=None, sigma=None, mode='soft',
...
         # The BayesShrink thresholds from [1]_ in docstring
         var = sigma**2
-        threshold = [{key: _bayes_thresh(level[key], var) for key in level}
-                     for level in dcoeffs]
+        if method == "BayesShrink":
+            # The BayesShrink thresholds from [1]_ in docstring
+            threshold = [{key: _bayes_thresh(level[key], var) for key in level}
+                         for level in dcoeffs]
+        else:
+            threshold = [{key: _universal_thresh(img, level[key], sigma)
```
Comment (grlee77) [inline skimage/restoration/_denoise.py]: There is no need to call `_universal_thresh` each time as the threshold is the same for all coefficient levels.  I would store the threshold in a variable and then build the dict using that single threshold

File: skimage/restoration/_denoise.py [hunk 3]
```diff
@@ -478,8 +490,8 @@ def _wavelet_threshold(img, wavelet, threshold=None, sigma=None, mode='soft',
     return pywt.waverecn(denoised_coeffs, wavelet)[original_extent]
 
 
-def denoise_wavelet(img, sigma=None, wavelet='db1', mode='soft',
-                    wavelet_levels=None, multichannel=False,
+def denoise_wavelet(img, sigma=None, method='BayesShrink', wavelet='db1',
+                    mode='soft', wavelet_levels=None, multichannel=False,
```
Comment (grlee77) [inline skimage/restoration/_denoise.py]: I think `method` should most likely be added as the last argument here to avoid breaking existing code.

**Comment by grlee77** [comment]
Please also add the following reference (section 4.2 is on VisuShrink)
```
DL Donoho and JM Johnstone, Ideal spatial adaptation by wavelet shrinkage, Biometrika, 1994, vol. 81, no. 3, pp. 425-455.
```

The implementation looks correct to me.  I just added a few style-related comments.  

In regards to sciunto's question, I would describe VisuShrink as having the specific property that as the signal size approaches infinity, the probability of any pure (Gaussian) noise coefficients surviving the threshold approaches zero (eq. 31 of the reference above).

Visually, the images produced will look overly smooth as compared to the BayesShrink algorithm.  There may be some instances in which the VisuShrink result is preferred, but the BayesShrink case is almost always going to give a result with higher MSE, PSNR and mSSIM metrics.

**Comment by monego** [comment]
Hi @sciunto 

> Is it still a work in progress?

The code itself is ready (needs a few adjustments as @grlee77 pointed) and the tasks in the message need to be worked on. Other than that, it is done.

> Could you explain in few words the differences between these two methods in practice, maybe with an example?

VisuShrink calculates the universal threshold based on the standard deviation of the diagonal detail coefficient, and being universal it is applied to all coefficients. This threshold is generally large and proportional to image size. The larger the image, more signal is killed along with noise: that leads to oversmoothing.

BayesShrink is subband adaptive and calculates a different threshold value for each detail coefficient. That means it has a better idea of what is noise and what is signal and denoising is more accurate to the original image.

I would show an image comparing both methods but apparently I need to upload it elsewhere then reference from Markdown.

**Comment by monego** [comment]
Hi @grlee77 

> why pass in details when it is not used?

I was locally testing a different threshold per subband and I forgot to remove the parameter before submitting the PR. Indeed, it doesn't make any difference here.

> There is no need to call `_universal_thresh` each time as the threshold is the same for all coefficient levels. I would store the threshold in a variable and then build the dict using that single threshold

> I think `method` should most likely be added as the last argument here to avoid breaking existing code.

> Please also add the following reference (section 4.2 is on VisuShrink)

Thanks for pointing this out. I will fix that.

> There may be some instances in which the VisuShrink result is preferred, but the BayesShrink case is almost always going to give a result with higher MSE, PSNR and mSSIM metrics.

While I can't think of any case where VisuShrink will give better results considering those metrics, it is a well known method commonly found in the literature and as a reference for comparing against other methods.

I read your suggestion on #1833  about implementing different thresholding methods. As you just said, while BayesShrink works fairly well on the general case, there may be special cases where others will be more suited. In page 2 of the article [1] referenced in the merged implementation it is said that BayesShrink a few times performed worse (though by only a 1% decrease of the MSE) than SUREShrink. NormalShrink [2] gives (slightly) better results than BayesShrink in several cases, and may offer a fair competition. It seems reasonable to offer some choices, especially when they are relatively easy to implement as only a small `_thresh` function and a condition block have to be added.

There are many other thresholding methods and adaptations of existing ones. If it's in the interest of skimage to extend the wavelet denoise functionality there should be a discussion of which to implement.

[1] Chang, S. Grace, Bin Yu, and Martin Vetterli. "Adaptive wavelet thresholding for image denoising and compression." IEEE Transactions on 9.9 (2000): 1532-1546. DOI: 10.1109/83.862633

[2] Kaur, L., Gupta, S., Chauhan, R. C., "Image Denoising using Wavelet Thresholding", Indian Conference on Computer Vision (2002), Graphics and Image Processing, Ahmedabad

**Comment by grlee77** [comment]
> While I can't think of any case where VisuShrink will give better results considering those metrics, it is a well known method commonly found in the literature and as a reference for comparing against other methods.

To be clear, I am in favor of providing VisuShrink as it is simple to implement and the universal threshold is very common in the literature!  BayesShrink is pretty good, but there are definitely some other methods that can outperform it.  One of the best I have evaluated is NeighShrink SURE [1], but it is not quite as simple in implementation as either VisuShrink or BayesShrink.

A simple way to substantially improve the performance of any of these (by around 1dB in PSNR) is to repeat the same algorithm with cycle-spinning [2] (i.e. average over multiple circular shifts to approximate a shift-invariant transform).

I have some code to do that, but haven't gotten around to making a PR for it.

Alternatively to cycle-spinning, the `swtn`, `iswtn` functions from `PyWavelets` could be used in place of `wavedecn`, `waverecn` to perform multi-level stationary (translation-invariant) wavelet transforms.  `iswtn` is currently only available via the master branch of `PyWavelets`, though, so that is not yet an option for `scikit-image`.

[1] Z. Dengwen, C. Wengang, "Image denoising with an optimal threshold and neighbouring window,"  Pattern Recognition Letters, vol.29, no.11, pp.1694–1697, 2008

[2] R. R. Coifman, D. L. Donoho.  Translation-Invariant De-Noising.  In:  Wavelets and Statistics.  Volume 103 of the series Lecture Notes in Statistics, pp. 125-150.

**Comment by sciunto** [comment]
@grlee77 So, would you recommend to keep only visushrink or put visushrink+Bayesshrink?

**Comment by grlee77** [comment]
In general BayesShrink should perform better than VisuShrink.  I would keep both and leave BayesShrink as the default.  Both are reasonably common in the literature, so it could be useful to have for comparison purposes.

**Comment by sciunto** [comment]
@monego would you like to apply the little changes requested by @grlee77?

**Comment by monego** [comment]
I would prefer that someone takes over. I am having technical problems and little time. Will try to find local guidance the next. Sorry for the inconvenience.

The docstring also needs some rewording (the `sigma` parameter assumes BayesShrink) and the VisuShrink threshold should fall into the (positive) isscalar test instead of building a separate dict.

**Comment by sciunto** [comment]
Thanks @monego 

@grlee77 I rebased the PR. Would you like to apply the changes the suggested? You can push to this branch directly.

**Comment by grlee77** [comment]
Yes, I can finish this up.  Thanks for the contribution @monego

**Comment by grlee77** [comment]
Please see the 3 most recent commits which address style issues and add an example and tests.

I rebased this on top of the bug fix in #2640.  Once that one has been merged I can do a final rebase here.

**Comment by monego** [comment]
That's much better, thanks. However, I would add:
* 'method' in `_wavelet_threshold` docstring can be copied from `denoise_wavelet` for consistency (for instance, in the latter it says 'The currently' and in the former 'Currently').
* 'method' in `denoise_wavelet` docstring is misplaced. It should be at the end.
* 'sigma' and 'threshold' are swapped in `_wavelet_denoise` docstring. That also occurs in the original/merged code.
* I added the word "currently" in case other methods (e.g. SURE) are to be added in the future. If VisuShrink and BayesShrink are the final choice then it's irrelevant to use that word. If SURE is to be added later, then that word could be kept, but it would probably not hurt to remove it either.
* Lastly, you added in `_wavelet_denoise` docstring, threshold parameter: "If it is set to None, a user-specified ``threshold`` must be supplied instead.", but is there a way for the user to supply one? I don't see how in the main function.

**Comment by monego** [comment]
Also in `denoise_wavelet` docstring, Notes, last paragraph:

> If ``mode == "VisuShrink"``

It should be method.

**Comment by grlee77** [comment]
>Lastly, you added in _wavelet_denoise docstring, threshold parameter: "If it is set to None, a user-specified threshold must be supplied instead.", but is there a way for the user to supply one? I don't see how in the main function.
No, there is not currently a way for a user to do this via the public `denoise_wavelet` function.  For some reason there is a unit test for that functionality, but perhaps it should either be removed or exposed as an additional optional `threshold` parameter in `denoise_wavelet`.  As `_wavelet_threshold` is a private function, I guess there would not really be a deprecation issue with just removing it.

**Comment by monego** [comment]
Ok, I got this now. I am going to ignore this manual threshold issue and leave it to be resolved in a separate PR.
I would say this PR for VisuShrink is now finished.

**Comment by monego** [comment]
Strings are now ok, style passes flake8, implementation is correct and the unit tests pass.

**Comment by grlee77** [comment]
thanks @monego.  I updated the expected warning string in one of the tests to match your change in e3a5acf

**Comment by soupault** [comment]
Thanks @monego @grlee77 !

**Pull request merged**